hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessorSimulator.java

```diff
@@ -0,0 +1,333 @@
+			mutableRelayer.decrementBalance(remainingFee);
+			allowanceCharged = remainingFee;
+		} else {
+			// If user gas price >= current gas price, sender pays all fees
+			final var senderCanAffordGas = mutableSender.getBalance().compareTo(gasCost) >= 0;
+			validateTrue(senderCanAffordGas, INSUFFICIENT_PAYER_BALANCE);
+			mutableSender.decrementBalance(gasCost);
+		}
+		// In any case, the sender must have sufficient balance to pay for any value sent
+		final var senderCanAffordValue = mutableSender.getBalance().compareTo(Wei.of(value)) >= 0;
+		validateTrue(senderCanAffordValue, INSUFFICIENT_PAYER_BALANCE);
+
+		var gasAvailable = gasLimit - intrinsicGas;
+
+		//construct the PrecompileMessage here
+		PrecompileMessage message = constructMessageAndCallPrecompileContract(sender, consensusTime, ledgers,
+				payload, gasAvailable, value, tokenId);
+
+		// calculate the gas used for the hts call
+		var gasUsedByTransaction = calculateGasUsedByTX(gasLimit, message.getGasRemaining());
+		final long sbhRefund = updater.getSbhRefund();
+
+		// return gas price to accounts
+		final long refunded = gasLimit - gasUsedByTransaction + sbhRefund;
+		final Wei refundedWei = Wei.of(refunded * gasPrice);
+
+		if (refundedWei.greaterThan(Wei.ZERO)) {
+			if (allowanceCharged.greaterThan(Wei.ZERO)) {
+				// If allowance has been charged, we always try to refund relayer first
+				if (refundedWei.greaterOrEqualThan(allowanceCharged)) {
+					mutableRelayer.incrementBalance(allowanceCharged);
+					mutableSender.incrementBalance(refundedWei.subtract(allowanceCharged));
+				} else {
+					mutableRelayer.incrementBalance(refundedWei);
+				}
+			} else {
+				mutableSender.incrementBalance(refundedWei);
+			}
+		}
+		// Send fees to coinbase
+		final long coinbaseFee = gasLimit - refunded;
+		MutableAccount mutableCoinbase = getMutableAccount(updater.getOrCreate(
+				Id.fromGrpcAccount(dynamicProperties.fundingAccount()).asEvmAddress()));
+		mutableCoinbase.incrementBalance(Wei.of(coinbaseFee * gasPrice));
+
+		if (dynamicProperties.shouldEnableTraceability()) {
+			stateChanges = updater.getFinalStateChanges();
+		} else {
+			stateChanges = Map.of();
+		}
+
+		updater.commit();
+
+		if (message.getState() == COMPLETED_SUCCESS) {
+			return TransactionProcessingResult.successful(
+					new ArrayList<>(),
+					gasUsedByTransaction,
+					sbhRefund,
+					gasPrice,
+					message.getHtsOutputResult(),
+					mirrorReceiver,
+					stateChanges);
+		} else {
+			return TransactionProcessingResult.failed(
+					gasUsedByTransaction,
+					sbhRefund,
+					gasPrice,
+					message.getRevertReason(),
+					exceptionalHaltReason,
+					stateChanges);
+		}
+
+	}
+
+	private PrecompileMessage constructMessageAndCallPrecompileContract(Account sender, Instant consensusTime, WorldLedgers ledgers,
+																		Bytes payload, long gasAvailable, long value,
+																		long tokenId) {
+		final var redirectBytes = constructRedirectBytes(payload, tokenId);
```

Comment:
Since we implement direct token calls, probably we can think of eliminating the use of redirect bytecode that we artificially build. If it's easily achievable we can try to directly invoke the correct ERCPrecompile (e.g. `balanceOf`, `totalSupply`, `transferFrom`). This would make even a better "shortcut" to execute the ERC precompile logic and reduce the complexity.